crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+///     QkSparseObservable *two = qk_obs_add(iden, iden);
+///
+///     double tol = 1e-6;
+///     QkSparseObservable *canonical = qk_obs_canonicalize(two);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_canonicalize(
+    obs: *const SparseObservable,
+    tol: f64, // no optional arguments in C -- welcome to the ancient past
+) -> *mut SparseObservable {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let result = obs.canonicalize(tol);
+    Box::into_raw(Box::new(result))
+}
+
+/// @ingroup QkSparseObservable
+/// Copy the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to a copy of the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *original = qk_obs_identity(100);
+///     QkSparseObservable *copied = qk_obs_copy(original);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_copy(obs: *const SparseObservable) -> *mut SparseObservable {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+
+    let copied = obs.clone();
+    Box::into_raw(Box::new(copied))
+}
+
+/// @ingroup QkSparseObservable
+/// Compare two observables for equality.
+///
+/// Note that this does not compare mathematical equality, but data equality. This means
+/// that two observables might represent the same observable but not compare as equal.
+///
+/// @param obs A pointer to one observable.
+/// @param other A pointer to another observable.
+///
+/// @return ``true`` if the observables are equal, ``false`` otherwise.
+///
+/// # Example
+///
+///     QkSparseObservable *observable = qk_obs_identity(100);
+///     QkSparseObservable *other = qk_obs_identity(100);
+///     bool are_equal = qk_obs_equal(observable, other);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` or ``other`` are not valid, non-null pointers to
+/// ``QkSparseObservable``\ s.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_equal(
+    obs: *const SparseObservable,
+    other: *const SparseObservable,
+) -> bool {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    let other = unsafe { const_ptr_as_ref(other) };
+
+    obs.eq(other)
+}
+
+/// @ingroup QkSparseObservable
+/// Print the observable.
+///
+/// @param obs A pointer to the ``SparseObservable`` to print.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     qk_obs_print(obs);
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_print(obs: *const SparseObservable) {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    println!("{:?}", obs);
+}
```

Comment:
We probably need to think a lot more about this interface.  I don't know how Rust's buffering wrapper around `Stdout` is going to interact with other things, like whatever the C program is doing with stdout, and we also probably need to consider how we define the interfaces here such that the output of this string-ification can be properly integrated with the slightly more general stream-based handlers, like `sprintf`, `fprintf` or potentially (in the future) C++'s stream operators.

Reply:
Yeah that's a good point, it doesn't interact well I think and things are not printed in order if one does
```c
qk_obs_print(obs);
printf("Some print statements are more equal than others");
```